### ISSUE TITLE
A few corrections on deal

### ISSUE THREAD

**Issue by orsinium** [beartype/beartype#35]
Hey, [deal](https://github.com/life4/deal) author here. Sorry for annoying you, it is your project and you can write anything you want there, I don't mind. However, I have a few small remarks on the [baddeal](https://github.com/beartype/beartype#baddeal) section:

1. deal is not intended to check types. See [Don’t check types](https://deal.readthedocs.io/details/recipes.html#don-t-check-types) section in the docs.
2. It's not a wrapper on top of typeguard. We use typeguard to check the return type of the tested function when using [deal.cases](https://deal.readthedocs.io/details/api.html#deal.cases). The flag `check_types` controls exactly this. This is the only place where typeguard is involved. I think I will even make it an optional dependency in a future release.
3. Deal has no open issue tracker but it is open for contributions, has a pretty decent test coverage (100% and mutable testing on top), and battle-tested on production for 3 years. Sure, it can be a deal-breaker for you (pun intended) but I don't think it negatively affects the project quality. At least, there were no bug reports for 2 years when the issue tracker was open (if I remember correctly).

I think both beartype and deal are great projects. However, I think they do different things and shouldn't be compared. Beartype checks type annotations at runtime, deal checks contracts at runtime, static time, generates tests, soonish will even have a [formal verifier](https://github.com/orsinium-labs/deal-solver) and a lot of other cool features that are unrelated to beartype. But as I said, the final decision is up to you.

**Comment by leycec** [comment]
> Hey, deal author here.

**Wow!** I'm deeply honoured. Deal looks phenomenal. I'm... kinda jealous, actually. You do so much that I'll just *never* have the time to address in @beartype. If you'd ever like to support O(1) runtime type-checking in deal via @beartype or *whatever*, I'd be delighted to have that on- or offline conversation. <sup>*do it do it do it*</sup>

> Sorry for annoying you...

You're not annoying *at all.* I welcome and cherish deep discussions like this. It's mostly why I'm still here on GitHub after all these long, *long* years.

Moreover, you're an extremely humble author. Your writing style reminds me quite a lot of self-effacing Japanese culture, which is my highest praise. I speak Japanese and attended  University in Japan, so... I just love that kind of communication style. :bow: 

> ...you can write anything you want there, I don't mind.

Oh, you're so magnanimous. That's incredibly refreshing. But... srsly. I'm of the formally rigorous academic mindset rather than the techbro "hype train" mindsight. So, I really want to get things certifiably and rigorously right rather than misrepresent either you or @beartype.

The superficial impression I get is that:

* Comparing @beartype and typeguard is fair and honest, because @beartype and typeguard **directly** compete in the same low-level runtime type-checking space.
* Conversely, comparing @beartype and deal is unfair and dishonest, because @beartype and deal only **indirectly** compete in the same low-level runtime type-checking space via typeguard.

That... makes perfect sense. You're absolutely right.

And even were that *not* the case (which it probably is), it's also unhelpfully redundant to directly tabulate deal in the [Runtime Type Checkers](https://github.com/beartype/beartype#id72) table. We already tabulate typeguard, which transitively includes (by extension) all typeguard-based technology stacks. Since tabulating all downstream Python packages that `import typeguard` would be insane, it's yet again unfair and dishonest that we only tabulate deal while omitting all of the *other* downstream Python packages that `import typeguard`.

So, we've concluded that I'm unfair and dishonest. :wink: 

> I think I will even make it an optional dependency in a future release.

This is actually a *great* idea, but for altogether different reasons than petty competitiveness on my part. Among one of the many filth-encrusted hats I wear is [Gentoo developer](https://github.com/leycec/raiagent), where I package *mostly* scientific software for Gentoo Linux. Since Gentoo is BallsCrazy™ by modern Linux standards, Gentoo packaging often proves... challenging.

Python packages that mostly only have optional dependencies are:

* Much easier to package on any Linux distro.
* Integrate much more cleanly with Gentoo Linux in particular, because we enable fine-grained control over installation of optional dependencies via [user-defined USE flags](https://wiki.gentoo.org/wiki/USE_flag).

I've actually been meaning to package deal for Gentoo as a larger push to package *everything* relating to Python type-checking. Shifting mandatory to optional dependencies would be a big personal help to me as a potential packager of your awesome software.

What I'm trying to say is this: :+1: :+1: :+1:

> Deal has no open issue tracker...

I'd... totally forgotten that. Having a closed issue tracker isn't *necessarily* a deal breaker, but it is a red flag as a fellow GitHubber. Of course, you probably have an excellent reason for shutting issues down. Unwanted malicious spam? Unhealthy stalker fandom? Whatever your reasons, they're valid. So, I get that.

From the community perspective, it'd still be great if you'd reconsider that decision at some point. Open-source as a whole really thrives on both issues *and* pull requests being openly readable and writeable to all participants. But you do! Deal is still ostensibly great stuff.

> has a pretty decent test coverage (100%...

By "pretty decent test coverage," most people just mean "at least 90% test coverage."

100% isn't pretty decent, Gram. That's **friggin' phenomenal test coverage.** Danish Gods! What I wouldn't give for 100% test coverage in @beartype. Just getting us to 95% broke my back. I shudder to think what going for 100% coverage would cost.

I only have so much sanity left.

> ...and mutable testing on top)

*Holy Danish pastries.* I didn't even know that [mutation testing](https://en.wikipedia.org/wiki/Mutation_testing) was a thing. Now that I do, I want that for @beartype. But nobody got time for that, so I'll have to depressingly settle for non-mutable 95% test coverage. *Curses!*

> ...battle-tested on production for 3 years.

This also impresses me. Beartype will *eventually* be used throughout our in-house closed-source suite of production web apps, but we're nowhere near there yet. You go, @orsinium!

> ...soonish will even have a formal verifier...

That... that... is *so* cool. Beartype wants one of those, too. :sob: 

> However, I think they do different things and shouldn't be compared.

You are correct, because you usually appear to be correct and this is no exception. I'll drop all reference to deal from our side-by-side profiling comparison.

Thanks for the detailed writeup, Gram. This was all fantastically helpful. I wish nothing but the utmost best for both you and the burgeoning deal community. Let's close this in a minute after I commit a hasty `README.rst` fix-up.

**Comment by orsinium** [comment]
Aww, thanks, I don't deserve it.

>  If you'd ever like to support O(1) runtime type-checking in deal via @beartype or whatever

That's an interesting idea. I'll give beartype a try and see if it makes sense to use it instead of (or in addition to) typeguard for optional type checking in `deal.cases`.

> Much easier to package on any Linux distro.

Yep, I know a bit about it. There were people who packaged [dephell](https://github.com/dephell/dephell) on different distros, and they also asked me to have fewer dependencies and stopped me from vendorizing them.

> From the community perspective, it'd still be great if you'd reconsider that decision at some point.

It would definitely be important for building community. However, my experience with [dephell](https://github.com/dephell/dephell) showed that I can't build a community (I didn't manage to find any other maintainer for the project to help me). And experience with [flakehell](https://github.com/life4/flakehell) showed that a project can attract the wrong community, I received a bunch of bad messages there, aggressively asking me to remove pieces of code, some features, and so on. Hence my careful approach, I don't want to be "these guys", I value your personal vision on the project and would never demand to remove anything from the readme or whatever.

After that experience, I make old-school open-source: without much feedback, without issue tracker, making the top priority my own fun and sanity. Still, I want to avoid making OSS for my reputation, I want it to be in some sense community projects, this is why it lives under @life4 (or @dephell in case of dephell ecosystem), not @orsinium.

> I didn't even know that mutation testing was a thing. Now that I do, I want that for @beartype.

[mutmut](https://pypi.org/project/mutmut/) is pretty good. Keep in mind that you need 100% test coverage first, otherwise it will report a lot of things from uncovered lines. And tests should be fast and idempotent. Nothing stops you from trying it right now, though, maybe it will find something interesting.

> Beartype will eventually be used throughout our in-house closed-source suite of production web apps, but we're nowhere near there yet.

I know that Zipsale uses deal internally. Their CTO has a talk about it (in Russian) and they recently released [flake8-functions-names](https://github.com/Melevir/flake8-functions-names/). Also, Nikita Sobolev a few years ago included it in Sant Petersburg's university IT course. And I use deal in my team, of course.

> That... that... is so cool. Beartype wants one of those, too

Type checking is a weak formal verification (and liquid types, refined types, or in my case contracts can make it a strong verification). So, in some sense, mypy already does the formal verification of types, no actions required from beartype.

Thank you for the warm welcome and understanding.

**Comment by leycec** [comment]
> I don't deserve it.

You definitely do. I feel deep commiseration with your negative dephell <sup>*best name ever*</sup> and flakehell <sup>*second best name ever*</sup> experiences. I'd never heard of either, but now wish I had. They both look gr8, m8 – especially dephell. Every Pythonista can relate to dependency hell, which means everyone wants dephell. Well, *wanted,* anyway.

What a shame, too! Nearly 2k GitHub stars is nothing short of an open-source miracle. Your product was clearly globally desired and consumed by a wide swath of developers. Did the scope just get out of hand a bit there? I can see how sanitizing Python packaging could rapidly metastasize into a shambolic maintenance nightmare capable of overwhelming even superstar lone-wolf devs such as yourself. :wolf: 

One specific aspect of dephell in particular would have been of *considerable* relevance to Gentoo Linux:

> Want to install a poetry based project, but don't like poetry? Just tell DepHell to convert the project's meta information into a setup.py and install it with pip.

We kinda hate `pyproject.toml` (and thus both `flit` and `poetry`) in Gentoo Land, because `poetry` itself is literally impossible to install with anything other than its own non-standard `get-poetry.py` script, which is bad because that script violates network sandboxing and thus sane packaging constraints. We ain't touching that with a ten-foot pole, so [lead Gentoo dev Michał Górny](https://github.com/mgorny) actually hand-rolled us [our own `pyproject2setuppy` Python package](https://github.com/mgorny/pyproject2setuppy) – which **does the exact same thing dephill did.**

**tl;dr:** Brilliant minds think alike.

> And experience with flakehell showed that a project can attract the wrong community...

Yikes! What a negative let-down that must have been. Flakehell looked so fun, too. I'm all over that 80's BBS-era ASCII art, which never fails to send me on a nostalgia-fueled trip down memory lane.

I have no good advice having not been there myself. Beartype users have been nothing but charitably awesome. <sup>...praise Thor</sup>

I *did* briefly attract this peculiar fellow who submitted unreadable non-issues and useless PRs that literally deleted everything on a [*very* unrelated repo](github.com/leycec/fsnrnue). After I kindly suggested we weren't having any of it and closed everything, he promptly disappeared back into the woodwork. His uniquely "word salad" approach to human communication suggested untreated mental illness, so it was hard to take anything personally. I mostly just felt bad.

Your case seems a whole lot more... personal. Ugh! Internet trolls are just the worst. What beautiful thing won't they ruin next? :disappointed: 

> After that experience, I make old-school open-source: without much feedback, without issue tracker, making the top priority my own fun and sanity.

That makes nothing but sense. Fun *has* to be the top priority, because it sure as {fecund-cow-patties} isn't the money!

You might find that `deal` attracts a more wholesome crowd this go around, though. Python packaging *is* hell, which might help explain why nobody was having fun there. Runtime type-checking and constraints validation, on the other hand? Pure, absolute, and unadulterated fun! Odin only knows why, but it just is.

I've tried explaining this to my wife. I fail every single time. We both accept this now.

> I'll give `beartype` a try and see if it makes sense to use it instead of (or in addition to) `typeguard` for optional type checking in `deal.cases`.

**You da man.** :+1: 

There's literally *no* overlap between `beartype` and `typeguard` at the moment, because the former only covers O(1) type-checking and the latter only O(n) type-checking. Since each have their respective tradeoffs, the growing `deal` userbase might appreciate being able to conveniently mix-and-match between the two to suite callable-specific needs.

We *do* hope to eventually infringe on `typeguard` territory with a decorator-based API enabling users to dynamically select the sort of performance guarantees they want on a per-callable basis (e.g., O(1), O(logn), O(n), or perhaps even hybrid strategies conditionally dependent on input size), but a distinct lack of funding means that's still a year or two away.

My kingdom for a sweet corporate sponsorship.

> Keep in mind that you need 100% test coverage first, otherwise it will report a lot of things from uncovered lines. And tests should be fast and idempotent.

* **Fast tests:** :heavy_check_mark: 
* **Idempotent tests:** :heavy_check_mark: 
* **100% test coverage:** *gods help me*

It may be some time measured in decades before `beartype` embraces `mutmut`. `</sigh>`

> I know that Zipsale uses deal internally. Their CTO has a talk about it (in Russian) and they recently released flake8-functions-names. Also, Nikita Sobolev a few years ago included it in Sant Petersburg's university IT course.

**Wow!** All of that is making me itch to package `deal` for Gentoo. If you haven't already, you might consider publicizing those references in a *"Used By"* (or something) section of your front-page documentation. It's all true and it's great advertisement, too.

> Thank you for the warm welcome and understanding.

Oh, thank you! I'll drop you a line when I eventually drop that Gentoo package for `deal`. Until then, have a great Summer and keep the fun open-source times a-rollin'.

**Comment by orsinium** [comment]
> so lead Gentoo dev Michał Górny actually hand-rolled us our own pyproject2setuppy Python package 

Long before dephell I had [poetry-setup](https://github.com/orsinium-archive/poetry-setup) and [install-requires](https://github.com/orsinium-archive/install-requires) for this.

> We kinda hate pyproject.toml (and thus both flit and poetry) in Gentoo Land

They both now support [PEP 517](https://www.python.org/dev/peps/pep-0517/) (provide a function to build wheel), and flit shouldn't need the network for it. And things are getting better. A few links you might find interesting:

+ https://github.com/takluyver/flit/issues/373
+ https://www.python.org/dev/peps/pep-0621/
+ https://www.python.org/dev/peps/pep-0631/
+ https://github.com/jazzband/pip-tools/pull/1356

> you might consider publicizing those references in a "Used By" (or something) section of your front-page documentation

Great idea, thank you.

Have a good summer too :) Feel free to [contact me](https://orsinium.dev/) anytime for any occasion.
